### PR TITLE
Docs/loading modules and data

### PR DESCRIPTION
**Pull request by mtomcal**
- Implemented documentation on "Loading Modules" and "Loading Data" sections.
- Linked API docs in Holocron for methods that belong to holocron to prevent duplication of effort.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -379,17 +411,101 @@ Documentation Forthcoming
...
+> 👍 Most commonly used method to load Holocron Modules
+
+A parent Module may add the `ModuleRoute` routing component to the [`childRoutes` Module Lifecycle Hook](#routing) to load a child Module dynamically on the server or browser when matching a route path. Once the Module is loaded, it is injected as a JSX element into the `children` prop of the parent Module.
+
+###### `ModuleRoute`
+
+<!-- ONE-MARKDOWN:START url="https://github.com/americanexpress/holocron/blob/master/packages/holocron/API.md" selector="### `ModuleRoute`" -->
+Please see [`ModuleRoute`](https://github.com/americanexpress/holocron/tree/master/packages/holocron-module-route#-usage) in the Holocron Module Route API.
+<!-- ONE-MARKDOWN:END -->
+
```
Comment (mtomcal) [inline README.md]: These sections are sparse as we decided to link to API methods that belong to other Libraries. I included a sample comment of how I envision the One Markdown module to work to "stitch" in API method definitions from other markdown files. If something goes wrong, it displays the contents of the "ONE-MARKDOWN:START" and "ONE-MARKDOWN:END" comment directives.

File: README.md [hunk 2]
```diff
@@ -379,17 +411,101 @@ Documentation Forthcoming
...
+| `store`   | [`Redux Store`](https://redux.js.org/api/store/) | Redux store containing `getState`, `dispatch` and [other methods](https://redux.js.org/api/store/). |
+| `fetchClient`   | [`fetch`](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch) | [ES6 Fetch API](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch) Compatible Client. |
+
+The `loadModuleData` Module Lifecycle Hook, is executed on the 1) Server and 2) Browser when a Module is loaded in either environment. This method is executed and resolved before any React Components are rendered inside a Holocron Module.
+
+In practice, we may [`dispatch`](https://redux.js.org/api/store/#dispatchaction) Redux actions and make [`async/await`](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Asynchronous/Async_await) requests to populate our Module's reducers before any React Components are rendered:
+
+```js
+// Runs on both Server and Browser
+HelloWorldModule.loadModuleData = async ({ store, fetchClient }) => {
```
Comment (anescobar1991) [inline README.md]: Can we link to where you can customize the `fetchClient`. I think we have that documented already right?

File: README.md [hunk 3]
```diff
@@ -379,17 +411,101 @@ Documentation Forthcoming
 
 ### Modules
 
-#### `loadModuleData`
+#### Loading Modules
 
-Documentation Forthcoming
+[Holocron Modules](#modules) are loaded and composed on 1) the Server when serving the HTML document and 2) in the Browser [Single Page App](https://developer.mozilla.org/en-US/docs/Glossary/SPA) mode via AJAX. We may load Holocron Modules when 1) a path is matched to the URL, or 2) dispatching a Redux action using an HOC and rendering with a React Component.
```
Comment (anescobar1991) [inline README.md]: I think we should simplify this sentence a bit as it is a bit confusing to read.  How about something like?

```
Holocron modules are loaded on 1). the server when serving the initial HTML and 2) in the Browser when the client JavaScript takes over and the [single page app](...) is initialized.

Holocron Modules are loaded either when a path is matched to a URL using the [child routes configuration](...), or when a React component uses [the Holocron API](...) to load and render a module.
```

File: README.md [hunk 4]
```diff
@@ -363,10 +363,42 @@ One App can be started via docker or if built from source by running `node lib/s
...
+HelloWorldModule.childRoutes = (store) => {
+  // return Route Components
+};
+
+// See "Loading Data" section
+HelloWorldModule.loadModuleData = async ({ store, fetchClient }) => {
+  // Async Requests on Module Load
+};
+
+// See "App Configuration" section
```
[the remark this answers is not in the document]
Comment (mtomcal) [reply README.md]: Not sure if you can link in code blocks with markdown....

[the remark this answers is not in the document]
Comment (anescobar1991) [reply README.md]: can you add the full url as a code comment?

**Comment by mtomcal** [comment]
- What is the communication plan for this change?
This is a communication plan for our API.
- Does any documentation need to be updated or added to account for this change? If so was it done already?
This is documentation
- What is the motivation for this change?
To document One App's API surface
- Should these changes also be applied to a maintenance branch or any other long lived branch?
No.
- How was this change tested?
N/A
- Does this change require cross browser checks? Why or why not?
N/A
- Does this change require a performance test prior to merging? Why or why not?
N/A
- Could this be considered a breaking change? Why or why not?
N/A
- Does the change impact caching?
N/A
- Does the change impact HTTP headers?
N/A
- Does the change have any new infrastructure requirements?
N/A
- Does the change affect other versions of the app?
N/A
- Does the change require additional environment variables?
N/A
- What is the impact to tenants?
N/A
- What is the impact to individual users?
N/A
- What is the change to the size of assets?
N/A
- Should integration tests be added to protect against future regressions on this change?
N/A

File: README.md [hunk 5]
```diff
@@ -379,17 +411,103 @@ Documentation Forthcoming
 
 ### Modules
 
-#### `loadModuleData`
+#### Loading Modules
 
-Documentation Forthcoming
+In both the Server and the Browser, the Holocron Modules loaded depends on which path is matched to a URL using the [child routes configuration](#routing), or when a React component uses the [Holocron API](https://github.com/americanexpress/holocron/blob/master/packages/holocron/API.md).
```
Comment (stephaniecoates) [inline README.md]: grammar nitpick: should be `the Holocron **Module** loaded depends` or `the Holocron Modules loaded **depend**`

Comment (stephaniecoates) [reply README.md]: based on the fact that a matched path _can_ load multiple modules, I think it should be `the Holocron Modules loaded depend`

File: README.md [hunk 6]
```diff
@@ -379,17 +411,103 @@ Documentation Forthcoming
...
+
+**Contents**
+* React Router Component
+  * [`ModuleRoute`](#moduleroute)
+* Load and Render
+  * [`RenderModule`](#rendermodule)
+  * [`holocronModule`](#holocronmodule)
+  * [`composeModules`](#composemodules)
+
+##### React Router Component
```
Comment (JAdshead) [inline README.md]: These are now `one-app-router` components. all instances of React Router should be One App Router

File: README.md [hunk 7]
```diff
@@ -379,17 +411,103 @@ Documentation Forthcoming
...
+
+Please see [`holocronModule`](https://github.com/americanexpress/holocron/blob/master/packages/holocron/API.md#holocronmodule) in the Holocron API.
+
+###### `composeModules`
+
+Please see [`composeModules`](https://github.com/americanexpress/holocron/blob/master/packages/holocron/API.md#composemodules) in the Holocron API.
+
+#### Loading Data
+
+When [Holocron Modules](#modules) are composed and loaded on the Server and Client, the `loadModuleData` Module Lifecycle Hook is called to load any async requests. The `fetchClient` injected into the `loadModuleData` Hook may be customized using [`createSsrFetch`](#createssrfetch).
```
Comment (JAdshead) [inline README.md]: The custom `fetchClient` is only for server side fetching with loadModuleData

Comment (mtomcal) [reply README.md]: This is not accurate as we 1) [load `fetchClient` with `enhancedFetch` clientside](https://github.com/americanexpress/one-app/blob/master/src/client/prerender.js#L35) and 2) [`composeModules` pulls `fetchClient` off of Thunk third argument](https://github.com/americanexpress/holocron/blob/master/packages/holocron/src/ducks/compose.js#L22).

File: README.md (hunk 5 above)
Comment (mtomcal) [reply README.md]: Fixed in fb577c4

File: README.md (hunk 7 above)
Comment (mtomcal) [reply README.md]: Fixed in fb577c4

File: README.md (hunk 6 above)
Comment (mtomcal) [reply README.md]: Fixed in fb577c4

File: README.md [hunk 8]
```diff
@@ -379,17 +411,103 @@ Documentation Forthcoming
...
+
+**Contents**
+* Route Component
+  * [`ModuleRoute`](#moduleroute)
+* Load and Render
+  * [`RenderModule`](#rendermodule)
+  * [`holocronModule`](#holocronmodule)
+  * [`composeModules`](#composemodules)
+
+##### React Router Component
```
Comment (anescobar1991) [inline README.md]: This still says `React Router`. Can you update to say 'One App Router Component`?

File: README.md [hunk 9]
```diff
@@ -379,17 +411,103 @@ Documentation Forthcoming
...
+
+##### `loadModuleData`
+
+**Runs On**
+* ✅ Server
+* ✅ Browser
+
+**Shape**
+```js
+HelloWorldModule.loadModuleData = async ({ store, fetchClient }) => {};
```
Comment (10xLaCroixDrinker) [inline README.md]: Don't we need to return a promise?

File: README.md (hunk 8 above)
Comment (mtomcal) [reply README.md]: Fixed in 7f9463d

File: README.md [hunk 10]
```diff
@@ -379,17 +411,103 @@ Documentation Forthcoming
...
+
+**Contents**
+* [`loadModuleData`](#loadmoduledata)
+* [`createSsrFetch`](#createssrfetch)
+
+##### `loadModuleData`
+
+**Runs On**
+* ✅ Server
+* ✅ Browser
```
[the remark this answers is not in the document]
Comment (mtomcal) [reply README.md]: I believe this convention will be used for other Module Lifecycle hooks. It is important to visually call out the fact that this is universal and I expect that `childRoutes` will have this metadata block in it as well. What do you think?

File: README.md (hunk 9 above)
Comment (mtomcal) [reply README.md]: Added `await Promise.resolve()`. Not sure if it muddies the waters though for concisely describing this function shape.

Comment (mtomcal) [reply README.md]: Fixed in b9da849

Comment (10xLaCroixDrinker) [reply README.md]: ya know what, I missed the `async` before 🤦‍♀️ lets revert this change 🙃

Comment (mtomcal) [reply README.md]: Reverted in db21fb46c23ce3c914fc2a5172ffd52ff97907aa

**Pull request merged**